kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/punctuators/AbstractThrottledPunctuator.java

```diff
@@ -0,0 +1,133 @@
+    long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
+    ArrayList<T> objectsAtWindow =
+        Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
+    boolean removed = objectsAtWindow.remove(object);
+    if (objectsAtWindow.isEmpty()) {
+      objectStore.delete(windowAlignedTimestamp);
+    } else {
+      objectStore.put(windowAlignedTimestamp, objectsAtWindow);
+    }
+    return removed;
+  }
+
+  @Override
+  public final void punctuate(long punctuateTimestamp) {
+    long startTimestamp = clock.millis();
+    log.debug(
+        "Processing tasks with throttling yield of {} until timestamp {}",
+        config.getYieldMs(),
+        punctuateTimestamp);
+    int keyCounter = 0;
+    int taskCounter = 0;
+    try (KeyValueIterator<Long, ArrayList<T>> it =
+        objectStore.range(getRangeStart(punctuateTimestamp), getRangeEnd(punctuateTimestamp))) {
+      // iterate through all keys in range until yield timeout is reached
+      while (it.hasNext() && canContinueProcessing(startTimestamp)) {
+        KeyValue<Long, ArrayList<T>> kv = it.next();
+        keyCounter++;
+        ArrayList<T> objects = kv.value;
+        long windowAlignedTimestamp = kv.key;
+        // collect all tasks to be rescheduled by key to perform bulk reschedules
+        Map<Long, ArrayList<T>> rescheduledTasks = new HashMap<>();
+        // loop through all objects for this key until yield timeout is reached
+        int i = 0;
+        for (; i < objects.size() && canContinueProcessing(startTimestamp); i++) {
+          T executingObject = objects.get(i);
+          taskCounter++;
+          TaskResult action = executeTask(punctuateTimestamp, executingObject);
+          action
+              .getRescheduleTimestamp()
+              .ifPresent(
+                  (rescheduleTimestamp) ->
+                      rescheduledTasks
+                          .computeIfAbsent(
+                              getWindowAlignedTimestamp(rescheduleTimestamp),
+                              (t) -> new ArrayList<>())
+                          .add(executingObject));
+        }
+        // process all reschedules
+        rescheduledTasks.forEach(
+            (timestamp, rescheduledObjects) -> {
+              ArrayList<T> mergedObjects =
+                  Optional.ofNullable(objectStore.get(timestamp)).orElse(new ArrayList<>());
+              mergedObjects.addAll(rescheduledObjects);
+              objectStore.put(timestamp, mergedObjects);
+            });
+        // all tasks till i-1 have been cancelled or rescheduled hence to be removed from store
+        if (i == objects.size()) {
+          // can directly delete key from store
+          objectStore.delete(windowAlignedTimestamp);
+        } else {
+          objectStore.put(
+              windowAlignedTimestamp, new ArrayList<>(objects.subList(i, objects.size())));
+        }
+      }
+    }
+    log.info(
```

Comment:
Change this to debug.

Reply:
@laxmanchekka preferred it to be info, similar in spirit to how kafka streams logs every update that it has processed x records and y punctuators